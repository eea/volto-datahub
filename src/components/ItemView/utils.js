const OBSOLETE_KEYS = ['obsolete', 'superseded'];

export const isObsolete = (status) => {
  let obsolete = false;
  if (status) {
    if (Array.isArray(status)) {
      if (
        status.filter((stat) => OBSOLETE_KEYS.includes(stat?.key)).length > 0
      ) {
        obsolete = true;
      }
    } else {
      if (OBSOLETE_KEYS.includes(status?.key)) {
        obsolete = true;
      }
    }
  }
  return obsolete;
};

export const isInternalURL = (url) => {
  let internal = false;
  url.split('/').forEach((part) => {
    if (part.split('_')[6] === 'i') {
      internal = true;
    }
  });
  return internal;
};

export const isInternal = (item) => {
  let internal = false;
  const links = item?.link || [];
  links.forEach((link) => {
    if (isInternalURL(link?.url)) {
      internal = true;
    }
  });
  return internal;
};

export const groupBy = (obj, key) => {
  return obj.reduce((rv, item) => {
    const group = rv[item[key]] || [];
    group.push(item);
    rv[item[key]] = group;
    return rv;
  }, {});
};

export const SVGIcon = ({ name, size, color, className, title, onClick }) => {
  return (
    <svg
      xmlns={name.attributes && name.attributes.xmlns}
      viewBox={name.attributes && name.attributes.viewBox}
      style={{ height: size, width: 'auto', fill: color || 'currentColor' }}
      className={className ? `icon ${className}` : 'icon'}
      onClick={onClick}
      dangerouslySetInnerHTML={{
        __html: title ? `<title>${title}</title>${name.content}` : name.content,
      }}
    />
  );
};
