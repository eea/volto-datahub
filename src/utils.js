const OBSOLETE_KEYS = ['obsolete', 'superseded'];

export const isInternalURL = (url) => {
  let internal = false;
  url.split('/').forEach((part) => {
    if (part.split('_')[6] === 'i') {
      internal = true;
    }
  });
  return internal;
};

export const isObsolete = (status) => {
  let obsolete = false;
  if (status) {
    if (Array.isArray(status)) {
      if (
        status.filter((stat) => OBSOLETE_KEYS.includes(stat?.key)).length > 0
      ) {
        obsolete = true;
      }
    } else {
      if (OBSOLETE_KEYS.includes(status?.key)) {
        obsolete = true;
      }
    }
  }
  return obsolete;
};

export const isInternal = (item) => {
  let internal = false;
  const links = item?.link || [];
  links.forEach((link) => {
    if (isInternalURL(link?.url)) {
      internal = true;
    }
  });
  return internal;
};

export const groupBy = (obj, key) => {
  return obj.reduce((rv, item) => {
    const group = rv[item[key]] || [];
    group.push(item);
    rv[item[key]] = group;
    return rv;
  }, {});
};

export function parseDatasets(children) {
  const all_datasets = children.map((v, e) => {
    let date;
    if (v.resourceTemporalExtentDetails) {
      const tc = v.resourceTemporalExtentDetails[0];
      const { start, end } = tc || {};
      const tc_start = start?.date
        ? new Date(Date.parse(start.date)).getFullYear()
        : '';
      const tc_end = end?.date
        ? new Date(Date.parse(end.date)).getFullYear()
        : '';
      if (tc_start === tc_end) {
        date = tc_start;
      } else {
        date = tc_start + '-' + tc_end;
      }
    } else {
      date = '';
    }

    return Object.assign({ temporalDateRange: date, ...v }, e);
  });

  const datasets = all_datasets.filter((v) => {
    return !(isObsolete(v.cl_status) || isInternal(v));
  });

  const archivedOrRestrictedDatasets = all_datasets.filter((v) => {
    return isObsolete(v.cl_status) || isInternal(v);
  });

  const groupDatasets = (items) => {
    const datasetsByCoverage = groupBy(items, 'temporalDateRange');

    const sortedDatasets = Object.keys(datasetsByCoverage).sort(
      function (a, b) {
        a = Math.max(...a.split('-'));
        b = Math.max(...b.split('-'));
        return b - a;
      },
    );

    const groupedDatasets = sortedDatasets.map((v, i) => {
      return { id: i, date: v, children: datasetsByCoverage[v] };
    });

    return groupedDatasets;
  };

  const groupedByTemporalCoverage = groupDatasets(datasets);
  const groupedByArchivedOrRestricted = groupDatasets(
    archivedOrRestrictedDatasets,
  );

  return {
    all_datasets,
    datasets,
    archivedOrRestrictedDatasets,
    groupedByTemporalCoverage,
    groupedByArchivedOrRestricted,
  };
}
