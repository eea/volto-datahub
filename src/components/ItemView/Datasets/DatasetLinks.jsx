import { Icon, List } from 'semantic-ui-react';
import { SVGIcon } from '../utils.js';

import lockSVG from 'remixicon/icons/System/lock-line.svg';
import { isInternalURL } from '@eeacms/volto-datahub/utils';

export default function DatasetLinks({ dataset }) {
  return (dataset['WWW:LINK'] || []).map((item, i) => {
    return (
      <List.Item key={i}>
        <List.Content>
          <div className="dataset-item">
            <Icon className="download" />
            {item.name ? (
              <a
                href={item.url}
                className="item-link"
                target="_blank"
                rel="noreferrer"
              >
                {item.name}
              </a>
            ) : (
              <a
                href={item.url}
                className="item-link"
                target="_blank"
                rel="noreferrer"
              >
                Download link
              </a>
            )}
            {isInternalURL(item.url) && <SVGIcon name={lockSVG} size="20px" />}
          </div>
        </List.Content>
      </List.Item>
    );
  });
}
