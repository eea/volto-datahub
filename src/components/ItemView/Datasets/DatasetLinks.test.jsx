import React from 'react';
import { render, screen } from '@testing-library/react';
import { isInternalURL } from '@eeacms/volto-datahub/utils';
import '@testing-library/jest-dom/extend-expect';

import DatasetLinks from './DatasetLinks';

jest.mock('remixicon/icons/System/lock-line.svg', () => 'MockLockIcon');
jest.mock('@eeacms/volto-datahub/utils', () => ({
  isInternalURL: jest.fn(),
}));

describe('DatasetLinks', () => {
  const dataset = {
    'WWW:LINK': [
      {
        url: 'https://example.com/link1',
        name: 'Link 1',
      },
      {
        url: 'https://example.com/link2',
        name: '',
      },
    ],
  };

  it('renders links from the dataset', () => {
    render(<DatasetLinks dataset={dataset} />);

    expect(screen.getByText('Link 1')).toBeInTheDocument();
    expect(screen.getByText('Download link')).toBeInTheDocument();
  });

  it('renders lock icon for internal URLs', () => {
    isInternalURL.mockReturnValue(true);

    const { container } = render(<DatasetLinks dataset={dataset} />);

    expect(container.querySelector('svg.icon')).toBeInTheDocument();
  });

  it('does not render lock icon for external URLs', () => {
    isInternalURL.mockReturnValue(false);

    const { container } = render(<DatasetLinks dataset={dataset} />);

    expect(container.querySelector('svg.icon')).not.toBeInTheDocument();
  });
});
