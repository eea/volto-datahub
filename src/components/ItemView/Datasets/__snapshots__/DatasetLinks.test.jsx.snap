// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DatasetEsri renders correctly 1`] = `
<div
  className="item"
  onClick={[Function]}
  role="listitem"
>
  <div
    className="content"
  >
    <div
      className="dataset-item"
    >
      <i
        aria-hidden="true"
        className="icon download"
        onClick={[Function]}
      />
      <a
        className="item-link"
        href="https://example.org/1"
        rel="noreferrer"
        target="_blank"
      >
        First
      </a>
    </div>
  </div>
</div>
`;
